File: src/main/java/org/cubexmc/metro/lifecycle/ScheduledTaskLifecycle.java
package org.cubexmc.metro.lifecycle;

import org.bukkit.Bukkit;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Minecart;
import org.bukkit.persistence.PersistentDataType;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.manager.LineManager;
import org.cubexmc.metro.manager.PortalManager;
import org.cubexmc.metro.manager.StopManager;
import org.cubexmc.metro.util.MetroConstants;
import org.cubexmc.metro.util.SchedulerUtil;
import org.cubexmc.metro.util.VersionUtil;

/**
 * Owns startup scheduled tasks that are not tied to a single listener.
 */
public class ScheduledTaskLifecycle {

    private final Metro plugin;
    private final LineManager lineManager;
    private final StopManager stopManager;
    private final PortalManager portalManager;
    private final TaskScheduler scheduler;
    private final boolean folia;
    private Object autoSaveTaskId;
    private Object legacyMinecartMigrationTaskId;

    public ScheduledTaskLifecycle(Metro plugin, LineManager lineManager, StopManager stopManager) {
        this(plugin, lineManager, stopManager, null);
    }

    public ScheduledTaskLifecycle(Metro plugin, LineManager lineManager, StopManager stopManager,
            PortalManager portalManager) {
        this(plugin, lineManager, stopManager, portalManager, new SchedulerUtilTaskScheduler(),
                VersionUtil.isFolia());
    }

    ScheduledTaskLifecycle(Metro plugin, LineManager lineManager, StopManager stopManager,
            PortalManager portalManager, TaskScheduler scheduler, boolean folia) {
        this.plugin = plugin;
        this.lineManager = lineManager;
        this.stopManager = stopManager;
        this.portalManager = portalManager;
        this.scheduler = scheduler;
        this.folia = folia;
    }

    public void start() {
        this.autoSaveTaskId = scheduler.schedule(plugin, this::processAsyncSaves, 1200L, 1200L);
        if (folia) {
            plugin.getLogger().info("Skipped legacy Metro minecart tag migration on Folia; full-world entity scans are not region-owned.");
            return;
        }
        this.legacyMinecartMigrationTaskId = scheduler.schedule(plugin, this::migrateLegacyMinecartTags, 100L, -1L);
    }

    public void shutdown() {
        if (autoSaveTaskId != null) {
            scheduler.cancel(autoSaveTaskId);
            autoSaveTaskId = null;
        }
        if (legacyMinecartMigrationTaskId != null) {
            scheduler.cancel(legacyMinecartMigrationTaskId);
            legacyMinecartMigrationTaskId = null;
        }
    }

    private void processAsyncSaves() {
        if (lineManager != null) {
            lineManager.processAsyncSave();
        }
        if (stopManager != null) {
            stopManager.processAsyncSave();
        }
        if (portalManager != null) {
            portalManager.processAsyncSave();
        }
    }

    private void migrateLegacyMinecartTags() {
        for (org.bukkit.World world : Bukkit.getWorlds()) {
            for (Entity entity : world.getEntitiesByClass(Minecart.class)) {
                if (MetroConstants.METRO_MINECART_NAME.equals(entity.getCustomName())
                        && !entity.getPersistentDataContainer().has(
                        MetroConstants.getMinecartKey(), PersistentDataType.BYTE)) {
                    entity.getPersistentDataContainer().set(
                            MetroConstants.getMinecartKey(), PersistentDataType.BYTE, (byte) 1);
                    plugin.getLogger().info("Migrated legacy Metro Minecart to PDC data: " + entity.getUniqueId());
                }
            }
        }
    }

    interface TaskScheduler {
        Object schedule(Metro plugin, Runnable task, long delay, long period);

        void cancel(Object taskId);
    }

    private static final class SchedulerUtilTaskScheduler implements TaskScheduler {
        @Override
        public Object schedule(Metro plugin, Runnable task, long delay, long period) {
            return SchedulerUtil.globalRun(plugin, task, delay, period);
        }

        @Override
        public void cancel(Object taskId) {
            SchedulerUtil.cancelTask(taskId);
        }
    }
}

File: src/main/java/org/cubexmc/metro/lifecycle/ScheduledTaskLifecycle.kt
package org.cubexmc.metro.lifecycle

import org.bukkit.Bukkit
import org.bukkit.entity.Minecart
import org.bukkit.persistence.PersistentDataType
import org.cubexmc.metro.Metro
import org.cubexmc.metro.manager.LineManager
import org.cubexmc.metro.manager.PortalManager
import org.cubexmc.metro.manager.StopManager
import org.cubexmc.metro.util.MetroConstants
import org.cubexmc.metro.util.SchedulerUtil
import org.cubexmc.metro.util.VersionUtil

/**
 * Owns startup scheduled tasks that are not tied to a single listener.
 */
class ScheduledTaskLifecycle {
    private val plugin: Metro
    private val lineManager: LineManager?
    private val stopManager: StopManager?
    private val portalManager: PortalManager?
    private val scheduler: TaskScheduler
    private val folia: Boolean
    private var autoSaveTaskId: Any? = null
    private var legacyMinecartMigrationTaskId: Any? = null

    constructor(plugin: Metro, lineManager: LineManager?, stopManager: StopManager?) : this(
        plugin,
        lineManager,
        stopManager,
        null,
    )

    constructor(
        plugin: Metro,
        lineManager: LineManager?,
        stopManager: StopManager?,
        portalManager: PortalManager?,
    ) : this(
        plugin,
        lineManager,
        stopManager,
        portalManager,
        SchedulerUtilTaskScheduler(),
        VersionUtil.isFolia(),
    )

    constructor(
        plugin: Metro,
        lineManager: LineManager?,
        stopManager: StopManager?,
        portalManager: PortalManager?,
        scheduler: TaskScheduler,
        folia: Boolean,
    ) {
        this.plugin = plugin
        this.lineManager = lineManager
        this.stopManager = stopManager
        this.portalManager = portalManager
        this.scheduler = scheduler
        this.folia = folia
    }

    fun start() {
        autoSaveTaskId = scheduler.schedule(plugin, Runnable { processAsyncSaves() }, 1200L, 1200L)
        if (folia) {
            plugin.logger.info("Skipped legacy Metro minecart tag migration on Folia; full-world entity scans are not region-owned.")
            return
        }
        legacyMinecartMigrationTaskId = scheduler.schedule(plugin, Runnable { migrateLegacyMinecartTags() }, 100L, -1L)
    }

    fun shutdown() {
        val autoSave = autoSaveTaskId
        if (autoSave != null) {
            scheduler.cancel(autoSave)
            autoSaveTaskId = null
        }
        val migration = legacyMinecartMigrationTaskId
        if (migration != null) {
            scheduler.cancel(migration)
            legacyMinecartMigrationTaskId = null
        }
    }

    private fun processAsyncSaves() {
        lineManager?.processAsyncSave()
        stopManager?.processAsyncSave()
        portalManager?.processAsyncSave()
    }

    private fun migrateLegacyMinecartTags() {
        val minecartKey = MetroConstants.getMinecartKey() ?: return
        for (world in Bukkit.getWorlds()) {
            for (entity in world.getEntitiesByClass(Minecart::class.java)) {
                if (
                    MetroConstants.METRO_MINECART_NAME == entity.customName &&
                    !entity.persistentDataContainer.has(minecartKey, PersistentDataType.BYTE)
                ) {
                    entity.persistentDataContainer.set(
                        minecartKey,
                        PersistentDataType.BYTE,
                        1.toByte(),
                    )
                    plugin.logger.info("Migrated legacy Metro Minecart to PDC data: " + entity.uniqueId)
                }
            }
        }
    }

    interface TaskScheduler {
        fun schedule(plugin: Metro, task: Runnable, delay: Long, period: Long): Any?

        fun cancel(taskId: Any)
    }

    private class SchedulerUtilTaskScheduler : TaskScheduler {
        override fun schedule(plugin: Metro, task: Runnable, delay: Long, period: Long): Any? =
            SchedulerUtil.globalRun(plugin, task, delay, period)

        override fun cancel(taskId: Any) {
            SchedulerUtil.cancelTask(taskId)
        }
    }
}
